src/basic/io-util.c

```diff
@@ -181,6 +182,29 @@ int ppoll_usec(struct pollfd *fds, size_t nfds, usec_t timeout) {
         return r;
 }
 
+int safe_ppoll_usec(struct pollfd *fds, size_t nfds, usec_t timeout) {
+        usec_t until;
+
+        assert(fds || nfds == 0);
+
+        until = usec_add(now(CLOCK_MONOTONIC), timeout);
+
+        for (;;) {
+                usec_t n;
+                int r;
+
+                r = ppoll_usec(fds, nfds, timeout);
+                if (!ERRNO_IS_TRANSIENT(r))
```

Comment:
Please check if `r` is negative before using `ERRNO_IS_FOO()`. So,
`if (r >= 0 || !ERRNO_IS_TRANSIENT(r))`

Reply:
Is this really necessary given the fact that we're testing the value returned by `ppoll_usec()` here which doesn't set `errno` in case of errors ?

Reply:
`ERRNO_IS_FOO()` internally calls `abs()`, hence positive return values can match. `ppoll_usec()` may returs `4` which equals to `EINTR`.